src/ln/keysinterface.rs

```diff
@@ -0,0 +1,186 @@
+use bitcoin::blockdata::opcodes;
+use bitcoin::network::constants::Network;
+use bitcoin::util::bip32::{ExtendedPrivKey, ExtendedPubKey, ChildNumber};
+
+use secp256k1::key::{SecretKey, PublicKey};
+use secp256k1::Secp256k1;
+use secp256k1;
+
+use ln::channel::ChannelKeys;
+use util::events;
+use util::logger::Logger;
+
+use std::sync::Arc;
+
+/// A trait to describe a wallet which sould receive data to be able to spend onchain outputs
+/// fron a lightning channel
+pub trait WalletInterface: Send + Sync {
+	/// Handle an incoming spendable_outputs message from SimpleManyChannelMonitor.
+	/// If SecretKey is None, wallet should use secret key of destination_script previously
+	/// provided to ChannelMonitor.
+	/// If SecretKey is Some, wallet should use it with basepoint_secret for local_delayed following
+	/// BOTL 3 derivation scheme.
+	fn handle_spendable_output(&self, event: events::Event);
+}
+
+/// Hacky custom output script descriptors to ease spending of onchain outputs by user wallet
+/// Maybe should be changed by real ones when merged into rust-bitcoin.
+/// StaticOutputs commit to a static pubkey, i.e one derived once for node operation lifetime.
+/// DynamicOutputs commit to a dynamic local_delayedpubkey, i.e one which change for each per_commitment_point
+pub enum CustomOutputScriptDescriptor {
+	/// Outpoint commits to a P2PWKH, should be spend by the following witness :
+	/// <signature> <pubkey> OP_CHECKSIG
+	StaticOutput {
+		/// Outpoint spendable by user wallet
+		outpoint: OutPoint,
+	},
+	/// Outpoint commits to a P2WSH, should be spend by the following witness :
+	/// <local_delayedsig> 0 <witnessScript>
+	/// With input nSequence set to_self_delay.
+	DynamicOutput {
+		/// Outpoint spendable by user wallet
+		outpoint: OutPoint,
+		/// local_delayedkey = delayed_payment_basepoint_secret + SHA256(per_commitment_point || delayed_payment_basepoint
+		local_delayedkey: SecretKey,
+		/// witness script encumbering output
+		witness_script: Script,
+		/// nSequence input must commit to self_delay to satisfy script's OP_CSV
+		to_self_delay: u16,
+	}
+}
+
+impl CustomOutputScriptDescriptor {
+	/// Build a StaticOuput descriptor
+	pub fn static_key(outpoint: OutPoint) -> Self {
+		CustomOutputScriptDescriptor::StaticOutput {
+			outpoint,
+		}
+	}
+
+	/// Build a DynamicOuput descriptor
+	pub fn dynamic_key(outpoint: OutPoint, local_delayedkey: SecretKey, witness_script: Script, to_self_delay: u16) -> Self {
+		CustomOutputScriptDescriptor::DynamicOutput {
+			outpoint,
+			local_delayedkey,
+			witness_script,
+			to_self_delay,
+		}
+	}
+}
+
+/// A trait to describe an object which should get secrets from user wallet and apply derivation
+/// to provide keys materials downstream
+pub trait KeysInterface: Send + Sync {
+	/// User wallet has to provide this KeysInterface with a master seed used to derive 
+	/// node_id as /0' and destination_script as /N'
+	fn provide_master_seed(&mut self, seed : &[u8; 32]);
```

Comment:
Why not just make the interface provide rust-lightning with keys and move the provide_master_seed into the KeysManager new()?

Reply:
Yes change, at first was thinking if user wallet want to change seed after init time, but contradicts with derivation happening only once

Reply:
Yes right, let the most of flexibility and see how users implement it